appsrc/dcfile/dctopgx.man
.TH DCTOPGX 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - DICOM image to PGX file"
.SH NAME
dctopgx \- ACR/NEMA DICOM PS3 ... DICOM PS3 - DICOM image to PGX file
.SH SYNOPSIS
.HP 10
.B dctopgx
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-quiet|silent
]
.so man1/optin.so
.so man1/binout.so
.SH DESCRIPTION
.LP
.B dctopgx
reads the named dicom or acr-nema single channel (grayscale or palette color; the photometric interpretation is not checked) input file and copies the raw image
pixel data to a PGX file as used for the JPEG 2000 Verification Model.
.LP
A PGX file contains of a single text header line of the form
"PG <byte order> [+|-]<bit-depth> <cols> <rows>\n", followed
immediately by the raw binary data. There may be 1, 2 or
4 bytes per sample corresponding to bit depths from 1 to 32.
Byte order is "LM" (little endian) or "ML" (big endian). The
signed or unsigned pixel representation is indicated by '+'
or '-' character before the bit depth. If the '+' is omitted,
unsigned is implied.
.LP
The input file must be unencapsulated.
.LP
PGX supports big or little endian
byte order in the case of > 8 bit data. The output byte order is always 
in the same as the input, since the raw bytes
are copied from input to output without any conversion. Use dccp(1)
first to change the encoding if necessary.
.LP
PGX also supports signed data, so the Pixel Representation of the
input is used to specify whether the output is signed or unsigned.
.LP
.SH OPTIONS
The PGX output goes where it is specified or standard out.
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-quiet|silent
.RS
Suppress the normal description of image parameters.
.RE
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR pnmtodc(1) ,
.BR dctopnm(1) ,
.BR rawtodc(1) ,
.BR dctoraw(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2022, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
230224:	libsrc/src/dctool/dcopto.cc, libsrc/src/dctool/elmdict.cc: Do not insert trailing null rather than space for private owners and private data elements with string VRs in copy/create/convert tools replace command line arguments (000582)

230121:	libsrc/standard/elmdict/dicom3.tpl: Retire the Robotic Base Location Indicator (CP 2245)

230119:	libsrc/standard/module/softcopy.tpl: Add ability to select optical path and use VOI and Palette Color in Advanced Blending Presentation State (CP 2036, CP 2037)

221230:	libsrc/standard/strval/ct.tpl: Extend list of Defined Terms for Rescale Type for classic CT Image to match current standard (Sup 158), and make error when not HU for non-LOCALIZER a warning (is enhanced requirement only and not mentioned in standard for classic)

221225:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Enhanced RT Image (Sup 213)

221224:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Support of new MLC types (CP 2229)

221123:	libsrc/standard/elmdict/dicom3.tpl: Correct Inventory Level,List of UID Matching Sequence keywords

221013:	libsrc/src/dctool/attrtypv.cc,v,locale/mesgtext.tpl: Check timezone in DT is correct length and form (000580) and  value content in DA, DT, TM even if incorrect length (000581)

221002:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Clinical Trial Time Point Type Code Sequence to Study Module (CP 2197)

220929:	libsrc/standard/elmdict/other.tpl: Add more UCSF BIRP private data elements from ACRIN 6698 ISPY2 Shared Private Tag Data Dictionary_20210520.xlsx (000579)

220923:	libsrc/standard/elmdict/dicom3.tpl: Correct data element tag for Maximum Number of Records that collides with Updated Metadata Sequence (Sup 223 FT3)

220016:	libsrc/standard/elmdict/other.tpl: Add UCSF BIRP private creator that have different case for XX than specified (000578)

220707:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Conceptual Volume in Structure Set (CP 2184)

220707:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for used ROI in Registration (CP 2183)

220630:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for waveform filter description (CP 2162)

220624:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for inventory supplement (Sup 223)

220612:	appsrc/dcfile/dccidump.cc: Add utility to dump Content Item Macro content not related to SR (000576)

220606:	libsrc/standard/strval/base.tpl: Add RRID coding scheme (CP 2164)

220606:	libsrc/standard/iodcomp,module/softcopy.tpl: Add profile IDCVOIPaletteOpticalPath to test draft CP adding VOI and Palette Color to Advanced Blending Presentation State as well as Specimen Module (000575)

220601:	libsrc/standard/condn.tpl,,sopcl.tpl,iodcomp,module,strval/softcopy.tpl: Add Advanced Blending Presentation State Storage IOD (Supp 189)

220524:	libsrc/standard/module/base.tpl: Add AcquisitionUID left out of General Acquisition Module when CP 2135 was added (001361)

220417:	libsrc/standard/strval/base.tpl: Update Coding Scheme UIDs and Designators to match current standard (2022b)

220412:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for Data Element Label Dictionary Module (ASTM E2339-21)

220409:	libsrc/standard/elmdict/other.tpl: Add private PixelMed data elements for original file name and TIFF IFD

220331:	libsrc/include/dctool/ie.h,src/dctool/ie.cc,standard/condn.tpl,elmdict/dicom3.tpl,iodcomp/base.tpl,dx.tpl,pet.tpl,rt.tpl,us.tpl,vl.tpl,xaxrf.tpl,module/base.tpl,vl.tpl,support/elmietoxml.awk: Add Acquisition and Pyramid Entities to Information Model, with UIDs and other Attributes (CP 2135); also add Annotation IE handling (Sup 222)

220331:	libsrc/standard/condn.tpl,module/base.tpl,vl.tpl: Relax requirement for Slice Thickness, Pixel Spacing and Imaged Volume Width/Depth/Height for Whole Slide Microscopy LABEL and OVERVIEW images (CP 2086)

220322:	libsrc/standard/condn.tpl,module/vl.tpl: Report missing Per-frame Functional Groups Sequence when WSI not TILED_FULL (000573)

220320:	libsrc/standard/elmdict/philips.tpl: Add unknown Philips MR presentation state private tag with VR encountered

220209:	libsrc/standard/elmdict/other.tpl: Add unknown Aperio private tag with VR encountered

220120:	libsrc/src/dctool/attrtypo.cc: Remove odd length assertion failure when reading OB so dataset read can continue; will detect later if verifying (000571)

220116:	libsrc/standard/elmdict/dicom3.tpl: CPACK 114 data elements (CP 2145, CP 2152, CP 2153, CP 2154, CP 2156, CP 2157)

211115: libsrc/support/condn.awk: Fix broken implementation of SequencePresentInPathFromRootFirstItemHasItems in conditions to correct spurious dciodvfy warning that ReferencedSeriesSequence should not be present references to instances in derivation per-frame functional group are present (000570)

211113:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Per-segment multiple algorithms and creators (CP 2115)

211109:	libsrc/standard/condn.tpl,module/rt.tpl: Add missing conditions for BlockThickness and BlockTransmission (000569)

211027:	libsrc/standard/elmdict/dicom3.tpl: Correct misspelled keyword for Number of Table Columns

211027:	libsrc/standard/strval/sr.tpl: Add UIDREF to Mammography CAD SR Value Types (CP 2053)

211026:	libsrc/standard/elmdict/dicom3.tpl: Correct capitalization of 'for' in keyword for (300A,0788)

211025:	libsrc/standard/strval/base.tpl,elmdict/other.tpl: Add 99_APERIO Coding Scheme Designator and private data elements per GT450 DX samples and DCS

211015:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check when UIDs are re-used for different entities, such as same for SOP Instance and Study (000568)

211009:	libsrc/include,src/dctool/ie.h,standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/vl.tpl: Add Microscopy Bulk Simple Annotations (Sup 222)

211005:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Tomotherapeutic and Robotic-Arm Treatment Modalities (Sup 176)


appsrc/dcfile/dctopgx.cc
static const char *CopyrightIdentifier(void) { return "@(#)dctopgx.cc Copyright (c) 1993-2022, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "mesgtext.h"
#include "bnopt.h"
#include "ioopt.h"
#include "dcopt.h"
#include "elmconst.h"
#include "transynd.h"
#include "transyn.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	BinaryOutputOptions	output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool quiet=options.get("quiet") || options.get("silent");

	dicom_input_options.done();
	output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	OutputOpenerFromOptions output_opener(
		options,output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< output_options.usage()
			<< " [-v|-verbose]"
			<< " [-quiet|-silent]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	ostream out(output_opener);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	Uint16 vRows = 0;
	Attribute *aRows = list[TagFromName(Rows)];
	if (!aRows)
		log << WMsgDC(MissingAttribute)
		    << " - \"Rows\""
		    << endl;
	else
		vRows=AttributeValue(aRows);

	Uint16 vColumns = 0;
	Attribute *aColumns = list[TagFromName(Columns)];
	if (!aColumns)
		log << WMsgDC(MissingAttribute)
		    << " - \"Columns\""
		    << endl;
	else
		vColumns=AttributeValue(aColumns);

	Uint16 vNumberOfFrames = 0;
	Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
	if (aNumberOfFrames)	// optional
		vNumberOfFrames=AttributeValue(aNumberOfFrames);

	char *vPhotometricInterpretation = 0;
	Attribute *aPhotometricInterpretation = list[TagFromName(PhotometricInterpretation)];
	if (!aPhotometricInterpretation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PhotometricInterpretation\""
		    << endl;
	else
		vPhotometricInterpretation=AttributeValue(aPhotometricInterpretation);

	Uint16 vSamplesPerPixel = 0;
	Attribute *aSamplesPerPixel = list[TagFromName(SamplesPerPixel)];
	if (!aSamplesPerPixel)
		log << WMsgDC(MissingAttribute)
		    << " - \"SamplesPerPixel\""
		    << endl;
	else
		vSamplesPerPixel=AttributeValue(aSamplesPerPixel);

	Uint16 vBitsAllocated = 0;
	Attribute *aBitsAllocated = list[TagFromName(BitsAllocated)];
	if (!aBitsAllocated)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsAllocated\""
		    << endl;
	else
		vBitsAllocated=AttributeValue(aBitsAllocated);

	Uint16 vBitsStored = 0;
	Attribute *aBitsStored = list[TagFromName(BitsStored)];
	if (!aBitsStored)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsStored\""
		    << endl;
	else
		vBitsStored=AttributeValue(aBitsStored);

	Uint16 vHighBit = 0;
	Attribute *aHighBit = list[TagFromName(HighBit)];
	if (!aHighBit)
		log << WMsgDC(MissingAttribute)
		    << " - \"HighBit\""
		    << endl;
	else
		vHighBit=AttributeValue(aHighBit);

	Uint16 vPixelRepresentation = 0xffff;
	Attribute *aPixelRepresentation = list[TagFromName(PixelRepresentation)];
	if (!aPixelRepresentation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PixelRepresentation\""
		    << endl;
	else
		vPixelRepresentation=AttributeValue(aPixelRepresentation);

	Uint16 vPlanarConfiguration = 0xffff;
	Attribute *aPlanarConfiguration = list[TagFromName(PlanarConfiguration)];
	if (vSamplesPerPixel > 1 && !aPlanarConfiguration)
		log << WMsgDC(MissingAttribute)
		    << " - \"PlanarConfiguration\""
		    << endl;
	else
		vPlanarConfiguration=AttributeValue(aPlanarConfiguration);

	if (!quiet) {
		log << "\tRows = " << dec << vRows << endl;
		log << "\tColumns = " << dec << vColumns << endl;
		log << "\tNumberOfFrames = " << dec << vNumberOfFrames << endl;
		log << "\tPhotometricInterpretation = "
		    << (vPhotometricInterpretation ? vPhotometricInterpretation : "")
		    << endl;
		log << "\tSamplesPerPixel = " << dec << vSamplesPerPixel << endl;
		log << "\tBitsAllocated = " << dec << vBitsAllocated << endl;
		log << "\tBitsStored = " << dec << vBitsStored << endl;
		log << "\tHighBit = " << dec << vHighBit << endl;
		log << "\tPixelRepresentation = " << dec << vPixelRepresentation << endl;
		log << "\tPlanarConfiguration = " << hex << vPlanarConfiguration << dec << endl;
		log << "\tRows*Columns*SamplesPerPixel*BitsAllocated/8 = "
		    << hex << (Uint32)vRows*vColumns*vSamplesPerPixel*vBitsAllocated/8
		    << dec << endl;
	}

	if (vNumberOfFrames > 1) {
		log << "Multiple frames not supported" << endl;
		success=false;
	}

	if (din.getTransferSyntaxToReadDataSet()->isEncapsulated()) {
		log << "Encapsulated (compressed) input data not supported" << endl;
		success=false;
	}

/*
       PGX is a trivial monochrome file format defined just
       to enable simple testing of JPEG2000 with image data having unusual
       bit-depths. The file consists of a single text header line of the form
       "PG <byte order> [+|-]<bit-depth> <cols> <rows>", with the binary data
       appearing immediately after the newline character, packed into 1, 2 or
       4 bytes per sample, depending upon the value of <bit-depth>. The <byte
       order> field must be one of the fixed strings, "LM" (LSB's first) or
       "ML" (MSB's first), while the optional `+' (default) or `-' character
       preceding the <bit-depth> field indicates whether the data is signed
       or not. Any bit-depth from 1 to 32 is acceptable.
*/

	const char *pgxstring=0;

	if (vSamplesPerPixel == 1) {
		pgxstring="PG";
	}
	else {
		log << "Only single channel images supported" << endl;
		success=false;
	}

	if (success) {
		Assert(pgxstring);
		out << pgxstring << " "
		    << (din.getEndian() == BigEndian ? "ML" : "LM") << " "
		    << (vPixelRepresentation > 0 ? "-" : " ")
		    << ((vBitsAllocated > 8 && vBitsStored < 8) ? 9 : vBitsStored) << " "
		    << vColumns << " " << vRows << "\n";
	}


	Attribute *apixeldata=list[TagFromName(PixelData)];
	if (!apixeldata) {
		log << EMsgDC(MissingAttribute) << " - \"PixelData\"" << endl;
		success=false;
	}
	else if (!apixeldata->isOtherData()) {
		log << EMsgDC(PixelDataIncorrectVR) << endl;
		success=false;
	}

	if (success) {
		OtherUnspecifiedLargeAttributeBase *
			opixeldata = apixeldata->castToOtherData();
		Assert(opixeldata);

		if (!quiet) log << "Writing ..." << endl;

		opixeldata->writeRaw(out);

		if (!out.good()) {
			log << EMsgDC(Writefailed) << endl;
			success=false;
		}
	}

	return success ? 0 : 1;
}

	
